src/Engine/GameEvents/HealUnitEvent.java

```diff
@@ -35,29 +35,19 @@ public void sendToListener(GameEventListener listener)
   @Override
   public void performEvent(MapMaster gameMap)
   {
+    int maxHP = unit.model.maxHP;
+
     if (null == payer)
-      unit.alterHP(repairPower);
-    else
+      unit.alterHP(repairPowerHP);
+    else if( unit.getPreciseHealth() < maxHP * 10 )
```

Comment:
This feels an awful lot like implementation-specific information living outside of `Unit`.

I don't like that health and HP are now distinct concepts; I should be able to compare `unit.getWhichever()` and compare that to `unit.model.maxHP` and have it be correct.
It might not hurt to also have a `unit.isHurt` just to avoid having to do a comparison.

Reply:
> This feels an awful lot like implementation-specific information living outside of `Unit`.

I suppose I could make a getter for "HP times 10", but I don't foresee this implementation detail ever changing.

> I don't like that health and HP are now distinct concepts; I should be able to compare `unit.getWhichever()` and compare that to `unit.model.maxHP` and have it be correct.

Right. That's why I had health stored as a floating point, initially.
Fundamentally, HP and health *aren't* the same thing. Health is just how the game tracks whether a unit should be dead or not.
HP is a related quantity that determines firepower scaling and such. I initially tried to get rid of HP as a separate concept, but I had to do a *lot* more hacking around in a similar vein to `maxHP * 10` (but more complicated, because I had to deal with rounding errors *everywhere*).
The example that comes to mind is `alterHP`. If `maxHP` is in range [1-100], then suddenly I can have a maxHP of 95... so I had to check the maximum bound *after rounding to HP*. I forgot I made the change and left it in. I think I'll back it out, since it's unnecessary.
It was a lot harder for me to keep track of what I was doing, since I didn't have clear terminology delineating what's in range [1-100] and what's in range [1-10], and it seems like most applications in the game logic care more about HP than health.

> It might not hurt to also have a `unit.isHurt` just to avoid having to do a comparison.

That's fair.

Reply:
> I don't foresee this implementation detail ever changing

Famous last words... especially given that you just changed it.

> then suddenly I can have a maxHP of 95

How? Because you are rounding to the nearest 1 instead of 10? I'm not sure I understand.

Reply:
> Famous last words... especially given that you just changed it.

It matches cart behavior as-is. They store health as [0-100] integers.
Seeing as this is a clone and exists to extend the original system, I don't think it's within scope for *this project* to change that.
If I'd had the certainty I have today a) about the implementation details and b) about floating point (in)consistency, I probably would have arrived at this solution as the least evil when I first wrote this stuff.

> How? Because you are rounding to the nearest 1 instead of 10? I'm not sure I understand.

That's exactly why.
`maxHP` is a public, non-final field of `UnitModel` for a reason. I don't want to support fractional HP in `maxHP`, as it makes no sense from the standpoint of the player.
I suppose I could also enforce that constraint via making a getter/setter for it, but I would much rather enforce it by datatype. I could make a `getMaxHealth()` if you want.

Reply:
This doesn't clarify how we end up with `maxHP==95`. We aren't setting different `maxHP`s for different units.

> I had to check the maximum bound after rounding to HP

Don't we convert to HP by doing a `ceil(health/10)`? No rounding involved. Certainly nothing that could end up going outside the valid HP range anyway.
I must be having a slow day again, because I am not tracking here.

Reply:
> This doesn't clarify how we end up with `maxHP==95`. We aren't setting different `maxHP`s for different units.

...yet.
I don't have specific plans to make a CO that changes maxHP, but I don't want to rule it out.

> Don't we convert to HP by doing a `ceil(health/10)`? No rounding involved. Certainly nothing that could end up going outside the valid HP range anyway.
> I must be having a slow day again, because I am not tracking here.

Okay. So.
In `alterHealth`, we add the current health and the change.
Then, if it is a positive alteration, we set the HP to its maximum value.

Let's say I have 81 health, with a cap of 95 health because my CO is a sadist.
A BBoat heals me.
That's +10 health, so I end up at 91 (or 10 HP)
Because that alteration was healing, we round up to 10 HP and re-assign health (this is cart behavior), and end up at 100 health.
Obviously, I must now apply the health cap and shift down to 95 health.
If, however, `maxHP` can only be an integer value, the rounding-up could not break anything so long as the HP cap had already been enforced.

Reply:
> but I don't want to rule it out.

... Why? damage values are determined based on a table; there's no reason to give a specific model more HP instead of just changing the damage it takes.

> ... because my CO is a sadist

I mean. I _guess_. But even still:

> Let's say I have 81 health...

This whole situation only arises due to converting to/from HP. If Health is the "native units" of the quantity, and we only convert to HP numbers for display, then there's no problem.

Reply:
> ... Why? damage values are determined based on a table; there's no reason to give a specific model more HP instead of just changing the damage it takes.

More HP isn't exactly the same as taking less damage, nor is it exactly the same as just having more defense.
One of the custom COs I've seen proposed has overheal on powers, which makes affected units extremely durable without changing their defense or damage chart... and can persist across turns.

> This whole situation only arises due to converting to/from HP. If Health is the "native units" of the quantity, and we only convert to HP numbers for display, then there's no problem.

That's a big "if", and one I will not be conceding any time soon.
HP is the "real" value. Health is used only to track when less-than-a-whole-HP damage should add up to a whole HP.

Reply:
> More HP isn't exactly the same as taking less damage, nor is it exactly the same as just having more defense.

True, but for separate unit types that doesn't really matter.

> One of the custom COs I've seen proposed has overheal on powers

Sure, but now we are in the land of trying to predict everything we could ever want, and I try to avoid YAGNI violations. I also try not to limit flexibility arbitrarily, but at some point you do need to have ground rules.

> HP is the "real" value.

Well, can we go the other way then? Store it as a [0-100] int in `Unit`, but nobody else ever gets to see anything but the [0-10] float value? I just feel like we are building a space lander with both a `getRangeToGround()` and `getAltitude()` which return meters and feet, or is it feet and meters, and we are gonna lose track and create a 5 billion dollar red dust cloud (though I guess the stakes are technically a bit lower here).
For example, `CombatEngine.calculateOneStrikeDamage()` doesn't specify which one it's returning, and I'm pretty sure it's not the same as what was available before.

Reply:
> True, but for separate unit types that doesn't really matter.

...what?

> Well, can we go the other way then? Store it as a [0-100] int in `Unit`, but nobody else ever gets to see anything but the [0-10] float value?

That's what I did initially, but you said "you're messing with stuff already, so do it right".
It's kind of a weird situation all around.

> I just feel like we are building a space lander with both a `getRangeToGround()` and `getAltitude()` which return meters and feet, or is it feet and meters, and we are gonna lose track and create a 5 billion dollar red dust cloud (though I guess the stakes are technically a bit lower here).

Yeah. That's why I've tried to be careful in labeling.

> For example, `CombatEngine.calculateOneStrikeDamage()` doesn't specify which one it's returning, and I'm pretty sure it's not the same as what was available before.

It specifies health. It's... unfortunate that "health" and "HP" are synonyms, but I don't have better terminology on me.

Reply:
> I'm just trying to avoid littering the `Unit` interface with excessive inscrutable accessors, and avoid multiple representations of the same value (DRY, etc).
> Right now we have:
> `getHP()`
> `getHPFactor()`
> `getPreciseHealth()`
> `getEffectiveHealth()`
> `effectiveHealth()` (static)
> `healthToHP()` (static)
> `healthFromHP()` (static)
> 
> This is a bit much, and some of these are only used in one or two places. I made a suggestion in the other [t]h[r]ead, so I won't bother repeating it here.

There's a balance to be struck between clean interfaces and duplicating logic, to be sure.
`getHP()` is probably the most important
`getHPFactor()` I made mostly just to de-duplicate code. I'm not gonna die on this hill either way
`getPreciseHealth()` is almost as important as `getHP()`, since it's the direct getter for `health`
`getEffectiveHealth()` is used for combat-calculation-adjacent things, where it needs to be compared to a damage number (We could re-standardize to a double for damage, but that means adding those rounding error adjustments back in)
`effectiveHealth()` acts as the DRY source for the round-up-to-HP operation on health
The last two are just converters.

Yeah, it seems like a bit much, I'll grant, but they all have a solid reason to exist.
To be fair, we would OBE most of them by going back to "always HP, sometimes floating point"... I just dunno that that's more desirable than what we have.

Reply:
> >    True, but for separate unit types that doesn't really matter.
> ...what?

I think I a brained my word or something. Nothing to see here.

> but you said "you're messing with stuff already, so do it right"

Immediately before that, I said:
> Maybe just keep it as the 0-100 percentage value at all times except when displaying it? That's closer to what I expected you meant when you talked about using integer calculations. The tricky bit there is that we'd have to get used to thinking in one scale despite seeing in another.

Which seems both very clear and very different than what happened. It would have been good to mention that you had a different fix in mind; then we could have pushed it to a new PR. On the other hand, if you think this implementation matches that description, then we are indeed speaking different languages.

> Yeah. That's why I've tried to be careful in labeling

Adding shades of meanings in a half-dozen overloads can quickly become more overwhelming than clear. Maybe you can keep them all straight, but I won't. I think it will be better if we can simplify the `Unit` interface to just the HP versions and only store it as the `health` value at rest.

> It specifies health. It's... unfortunate that "health" and "HP" are synonyms, but I don't have better terminology on me.

So let's consolidate. Please. I don't care if we call it FluffyTamarin, I just want it to be one thing, at least outside of `Unit`. Maybe even call it `healthPercent` inside `Unit`, since that's what you said it really is (We are talking about allowing different `maxHP` values, not different `health` values, right?).

Reply:
> "always HP, sometimes floating point"... I just dunno that that's more desirable than what we have.

I will vote in the affirmative.

Reply:
> Which seems both very clear and very different than what happened. It would have been good to mention that you had a different fix in mind; then we could have pushed it to a new PR. On the other hand, if you think this implementation matches that description, then we are indeed speaking different languages.

Well. I *started* to do what you suggested, then said "wait, that's really hard to keep track of" because the important thing is HP.

> > "always HP, sometimes floating point"... I just dunno that that's more desirable than what we have.
> 
> I will vote in the affirmative.

Okay. Doesn't thrill me, but I'll live.